### ISSUE TITLE
Editor - Ability to detect when a screen reader is running

### ISSUE THREAD

**Issue by Lanie-Molinar** [freeCodeCamp/freeCodeCamp#42396]
I'm a blind user of FreeCodeCamp. I love that the editor you use has accessibility support, but as of now, you have to turn on screen reader optimization manually each time you use the editor. This can get quite tedious when you're switching between the info and code tabs on challenges, or completing easier challenges quickly.
When I looked at the help documentation for the editor you use, the guide for developers suggests to automatically detect if a screen reader is running, which will allow the editor to automatically turn on accessibility support. That guide can be found [here](https://github.com/microsoft/monaco-editor/wiki/Accessibility-Guide-for-Integrators).
The relevant portion of the guide is quoted below:

> 5. Do you have a way to detect screen readers on your runtime / Do you have a user setting for it?
> Accessibility best practices dictate that applications should be accessible by default, and that no special behaviour should be implemented for accessibility purposes. I would invite the authors of the best practices to try to write a Screen Reader friendly code editor using w3c's ARIA :).
> Anyways, the state of the art is lacking. Here's an example: there is no way to communicate editor wrapping points to Screen Readers, and they will quickly get confused when pressing for example arrow down in a wrapped line; they will expect the system caret to move to the next line, and we will move the system caret to the same model line, at a different column (since the line is wrapped); so, when we know a Screen Reader is connected to an editor, we disable wrapping.
> If you have a user setting (like e.g. Google Docs) or run in an environment where it is possible to find out if a Screen Reader is attached (like e.g. Electron or a WebBrowser control), then please configure the editor with accessibilitySupport: "on". Even if you don't do this, your users will be let know in the editor's aria label that they can optimize the editor by themselves using Alt+F1 (or Ctrl+F1 in IE) and following the directions in there.

Can such a user setting be added so that accessibility support will be enabled automatically? Failing that, can you add a way for the editor to remember the setting for accessibility support once you toggle it?

**Comment by raisedadead** [comment]
Hi Lanie,

Thanks for taking the time to report this and going further to investigate and find us the docs to implement this. Let us take a look and get back to you.

Thanks.

**Comment by huyenltnguyen** [comment]
I see that we have a `inAccessibilityMode` state stored in redux, which is reset when the page is refresh. 

I'm wondering if saving it to local storage would be enough to have it persisted, rather than implementing a screen reader detection of some sort. It's the privacy implications that I don't feel comfortable about.

**Comment by raisedadead** [comment]
Local storage should be fine, I think.

> It's the privacy implications that I don't feel comfortable about.

As for detecting the screen reader, We should be clear as long as we never collect this information outside of the client. The idea is to be able to assist as much as we can. 

If it's stored and used on the client only, it should be OK. We could ask for consent with an alert or modal when we detect and use the screen reader persistence.

**Comment by huyenltnguyen** [comment]
Hi @Living4God1991, I'm looking at the issue right now. I will have the setting saved to the browser local storage, which helps prevent it from being automatically reset. 

Though, I'm wondering if you are aware of the Ctrl + F1 (or Cmd + F1 on Mac) keyboard combination. The combination behaves like a toggle that switches the accessibility support on or off, and saves the setting until the page is refreshed. This is not ideal, but since moving between lessons does not trigger page reload, it will help reduce the annoyance a little.

The reason I ask is because I found a couple of UX/accessibility issues when testing the editor behavior, and I'm attempting to address them as well:
- **List of keyboard shortcuts is not present on the page.** I think it will be helpful for both visual and non-visual users to have it displayed. In fact, I could only find the keyboard combination above when I dug into the code base, and I assume other users wouldn't know about these. My suggestion is adding a button to the button group, maybe under the "Get Help" one, and show a dialog when the user clicks on it. 
- **Lack of user feedback when a keyboard shortcut is pressed.** I noticed this as I was testing if the Ctrl + F1 combination is working. I had to open the dev tool to verify because there wasn't any visual changes. We should show a notification of some sort when the user triggers an action, and must ensure that the message is picked up by screen readers. 

If you find other accessibility issues, please let us know.

**Comment by Living4God1991** [comment]
Hi @huyenltnguyen. I had no idea this keyboard shortcut existed. I've started using it for now. It would be great to have something showing the available keyboard shortcuts. I also noticed the lack of feedback when a keyboard shortcut is pressed. After I pressed CTRL+F1, I still found myself pressing Alt+F1 to see if the editor was optimized for screen readers.

**Comment by ojeytonwilliams** [comment]
We're planning to make the keyboard shortcuts more obvious for the upcoming curriculum, but we don't have a firm timeline for that.  In the short term, would it be helpful to have a bit of text for screen readers that tells you when the accessibility mode is off and how to enable it? Something like "Accessibility support off, press Ctrl + F1 (or Cmd + F1 on Mac) to enable"

**Comment by huyenltnguyen** [comment]
@ojeytonwilliams Sorry, but I kind of want to fight for the presence of keyboard shortcuts on the page 😅 

It is not just for the Ctrl + F1 combination alone, and nor for screen readers support specifically, but I believe visual users can get the benefits as well. There was a user asked about the Ctrl + Enter shortcut the other day in the chat because they couldn't find it anywhere on the screen.

Unless there is an overhead cost that I'm overlooking, my proposal is to add another button under the Get Help one, and show the list of keyboard shortcuts in a dialog when the button is pressed. And this a quick demo I made:

![keyboard-shortcuts](https://user-images.githubusercontent.com/25715018/122033738-06d46400-cdfb-11eb-879a-b80c41629b3c.gif)

**Comment by ojeytonwilliams** [comment]
> @ojeytonwilliams Sorry, but I kind of want to fight for the presence of keyboard shortcuts on the page 

Yep, I understand.  I broadly agree, too.  The only issue (to my knowledge) is that we don't want to advertise the navigation hotkeys too heavily.  Reason being, they're really useful for us as contributors, but we don't want to encourage people to skip challenges.  If we only 'advertise' the other shortcuts I think that should be fine.

Also, it might be better to put them as another option under 'Get Help' to save space on the screen.

cc: @ahmadabdolsaheb

**Comment by Nirajn2311** [comment]
I feel that adding it in 'Get Help' might not make it known to the users that we have an option to view the list of keyboard shortcuts. Maybe it could be something similar to how it is in coderadio(but this might mean accessibility issues) or it could be added to the navbar but only in challenges.

**Comment by ahmadabdolsaheb** [comment]
We will have the multiple editors for different file formats (css, html, js) in the upcoming UI iterations. When all of the editors are unselected, the user will see the list of shortcuts that includes search, challenge, navigation, and accessibility shortcuts. 

Most of the shortcuts are hidden in that way for the reasons Oliver mentioned above.

Since The challenge UI is the most visited UI, we are restrained to only allocate real estate to info and functionally that is essential for solving each challenge. The UI gets a bit more crammed on mobile; therefore adding a shortcut button that may only be used a couple of times is less justifiable.

That being said, I understand how having that button could be helpful for non-visual users to toggle the accessibility support.

Would a toggle in the settings that persists in the localhost be an option? We could add the shortcut info below the toggle similar to the additional info we provide to users in privacy settings.

**Comment by huyenltnguyen** [comment]
You know, it would be strange if I don't come back to an issue discussion with a couple more issues. 

I believe that the actions we add to [Editor.js](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Challenges/classic/Editor.js#L263) should show up on the accessibility support dialog when we press Alt + F1, but they don't seem to show up:

<img width="587" alt="Screen shot of the accessibility dialog" src="https://user-images.githubusercontent.com/25715018/122254209-19cc5e80-cef7-11eb-82c3-ee206282fbe0.png">

As you can see in the screenshot, VSCode editor has the Ctrl + E (Cmd + E on Mac) combination built-in to toggle the screen reader optimization, which is exactly what Ctrl + F1 does on our side. The thing is, only the Ctrl + F1 is bound to a Redux action, which saves the setting to Redux store. Since Ctrl + F1 is not listed in the keyboard shortcut list, users will follow the default instruction and use Ctrl + E instead, and the setting gets reset every time the editor unmounts.

I wonder if it would be simpler if we change our code to listen to the Ctrl + E combination instead, and let VSCode handle the announcement for us when the state changes. See their source code for it [here](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/codeEditor/browser/accessibility/accessibility.ts#L196-L225).

Another issue is, as Lanie pointed out, there is currently no quick way to know if the feature is on or off, as in the user would need to focus onto the editor and press Alt + F1 to check it. I still need to figure out if this is an issue on our side or on VSCode side, though we can do something with our focus event handler obviously.

---

Regarding the display of keyboard shortcut list, I have raised a draft PR (https://github.com/freeCodeCamp/freeCodeCamp/pull/42516), but it is not complete and is more of a proof-of-concept at the moment. 

What I didn't agree with initially was that leaving the keyboard shortcuts, especially those that are helpful for screen reader users, completely inaccessible / hard to discover. But that was before I look closely at the editor implementation and found the dialog issue above. 

I have 3 keyboard shortcuts listed in my PR and I don't plan to add more. They are:
1. Ctrl + F1
2. Ctrl + Enter
3. Ctrl + S

IMHO, (1) and (2) are must have to me accessibility-wise; and (3) is good-to-have. If we are able to communicate with screen readers about (1) and (2), the keyboard shortcut list will just be a UX issue, and can be discussed and addressed separately. 

---

So for the accessibility issues at hand, the minimum resolution I would suggest is:
- Use Ctrl + E keyboard shortcut in place of Ctrl + F1
- Save user setting to local storage on Ctrl + E press
- Announce the state of the setting on editor focus
- Use [`aria-describedby`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute) on the "Run the Tests" button to inform screen reader users about its shortcut

**Comment by Nirajn2311** [comment]
Ummm isn't Ctrl+S the browser keybind to save a webpage. Not like many people use it, but should we really be tampering/changing the logic of that.

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen, thanks for doing further research. That makes sense to me. We could communicate the state of accessibility mode if necessary on the editor non-visually so the screen readers would pick it up.
